### PR TITLE
Update the code of APT service(part2)

### PR DESCRIPTION
**Pull request by JamePeng**
The part2 update some stub codes 
Otherwise, I'd like to research something！

### CODE REVIEW

File: src/core/hle/applets/applet.cpp [hunk 1]
```diff
@@ -68,6 +68,17 @@ std::shared_ptr<Applet> Applet::Get(Service::APT::AppletId id) {
...
 }
 
+u32 GetRegisteredAppletCount() {
+    u32 registered_count = 0;
+
+    for (auto itr = applets.begin(); itr != applets.end(); ++itr) {
+        if (itr->second != nullptr) {
+            ++registered_count;
+        }
+    }
```
Comment (wwylele) [inline src/core/hle/applets/applet.cpp]: Can be simplified using `std::count_if`

File: src/core/hle/service/apt/apt.cpp [hunk 2]
```diff
@@ -41,6 +45,10 @@ static std::shared_ptr<std::vector<u8>> shared_font;
 
 static u32 cpu_percent; ///< CPU time available to the running application
 
+static AppletPreparationState applet_preparation_state = AppletPreparationState::NoPreparation;
+static CaptureBufferInfo capture_buffer_info;
+static WirelessRebootInfo wireless_reboot_info;
```
Comment (Subv) [inline src/core/hle/service/apt/apt.cpp]: Note for the future, I think these variables (and all other APT state variables) are stored in a per-session array, as in, one storage location for each open session to the APT service. This should be researched more.

File: src/core/hle/service/apt/apt.cpp [hunk 3]
```diff
@@ -272,36 +471,87 @@ void PrepareToStartApplication(Service::Interface* self) {
...
+     * The utility SleepIfShellClosed deside the 3DS system whether in sleep state at short intervals (by check PTM::shell state?)
+     */
+    case UtilityID::SleepIfShellClosed:
+        shell_open = PTM::GetShellState();
+        if (!shell_open) {
+            LOG_WARNING(Service_APT, "(STUBBED) SleepIfShellClosed, The 3DS System should sleep now! :p");
+        } else {
+            LOG_WARNING(Service_APT, "(STUBBED) SleepIfShellClosed shell_open=%u",shell_open);
+        }
+        break            ;
```
Comment (Subv) [inline src/core/hle/service/apt/apt.cpp]: What's up with those whitespaces?

File: src/core/hle/service/apt/apt.cpp [hunk 4]
```diff
@@ -93,11 +101,176 @@ void GetSharedFont(Service::Interface* self) {
...
+
+        mainscreen_color_per_pixel = GPU::Regs::BytesPerPixel(static_cast<GPU::Regs::PixelFormat>(capture_buffer_info.mainscreen_display_buffer_mode));
+        subscreen_color_per_pixel = GPU::Regs::BytesPerPixel(static_cast<GPU::Regs::PixelFormat>(capture_buffer_info.subscreen_display_buffer_mode));
+
+        capture_buffer_info.size = 240 * (mainscreen_color_per_pixel * 400 + subscreen_color_per_pixel * 320);
+
+        memcpy(Memory::GetPointer(buffer_addr), &capture_buffer_info, sizeof(CaptureBufferInfo));
+
+        cmd_buff[1] = RESULT_SUCCESS.raw; // No error
+        LOG_WARNING(Service_APT, "(STUBBED) buffer_addr=0x%08X, buffer_size=0x%08X", buffer_addr, buffer_size);
```
Comment (Subv) [inline src/core/hle/service/apt/apt.cpp]: Is this really a stub?

File: src/core/hle/service/apt/apt.cpp [hunk 5]
```diff
@@ -93,11 +101,176 @@ void GetSharedFont(Service::Interface* self) {
...
+        cmd_buff[2] = buffer_size;
+        LOG_ERROR(Service_APT, "(STUBBED)The buffer size is invalid! buffer_size=0x%08X", buffer_size);
+        return;
+    }
+
+    memcpy(&capture_buffer_info, Memory::GetPointer(buffer_addr), sizeof(buffer_size));
+
+    cmd_buff[1] = RESULT_SUCCESS.raw; // No error
+    cmd_buff[2] = buffer_size;
+    LOG_WARNING(Service_APT, "(STUBBED) buffer_size=0x%08X", buffer_size);
```
Comment (Subv) [inline src/core/hle/service/apt/apt.cpp]: Is this really a stub?

File: src/core/hle/service/apt/apt.cpp (hunk 2 above)
Comment (JamePeng) [inline src/core/hle/service/apt/apt.cpp]: yes！I think so！

Comment (JamePeng) [inline src/core/hle/service/apt/apt.cpp]: @Subv what content should I note？

File: src/core/hle/service/apt/apt.h [hunk 6]
```diff
@@ -31,6 +33,26 @@ struct AppletStartupParameter {
...
+    u32 mainscreen_display_buffer_mode;
+    u32 subscreen_left_offset;
+    u32 subscreen_right_offset;
+    u32 subscreen_display_buffer_mode;
+};
+static_assert(sizeof(CaptureBufferInfo) == 0x20, "CaptureBufferInfo struct has incorrect size");
+
+struct WirelessRebootInfo {
+    std::array<u8, 6> mac_address = { 0, 0, 0, 0, 0, 0 };
+    std::array<u8, 9> wireless_reboot_passphrase = { 0, 0, 0, 0, 0, 0, 0, 0, 0 };
```
Comment (lioncash) [inline src/core/hle/service/apt/apt.h]:
```cpp
std::array<u8, 6> mac_address{};
std::array<u8, 9> wireless_reboot_passphrase{};
```

does the same thing.

File: src/core/hle/service/apt/apt.h [hunk 7]
```diff
@@ -73,6 +108,48 @@ enum class StartupArgumentType : u32 {
     OtherMedia = 2,
 };
 
+enum class QueryReply : u32 {
+    Reject = 0,
+    Accept = 1,
+    Later  = 2
+};
+
+enum class AppletPreparationState :u32 {
```
Comment (lioncash) [inline src/core/hle/service/apt/apt.h]: space after the colon

File: src/core/hle/service/apt/apt.cpp (hunk 2 above)
Comment (Subv) [inline src/core/hle/service/apt/apt.cpp]: That's orthogonal to this PR, it was just an observation

File: src/core/hle/service/apt/apt.h [hunk 8]
```diff
@@ -262,32 +495,35 @@ void CancelParameter(Service::Interface* self);
...
- *     3 : u8 flag
- *     4 : (Size0<<14) | 2
- *     5 : Buffer 0 pointer
- *     6 : (Size1<<14) | 0x802
- *     7 : Buffer 1 pointer
+ *     1 : Parameter Size (capped to 0x300)
+ *     2 : HMAC Size (capped to 0x20)
+ *     3 : u8 Paused (0 = not paused, 1 = paused)
+ *     5 : void*, Parameter
+ *     7 : void*, HMAC
```
Comment (wwylele) [inline src/core/hle/service/apt/apt.h]: why remove input 4 and 6

Comment (JamePeng) [inline src/core/hle/service/apt/apt.h]: @wwylele Otherwise,do you know the mean of AppletUtility::Transition?

Comment (wwylele) [inline src/core/hle/service/apt/apt.h]: @JamesPeng sorry, not really... OK I won't say more since I don't know much about APT service.

Comment (Subv) [inline src/core/hle/service/apt/apt.h]: Still, you shouldn't remove them, they are descriptors for the buffers

File: src/core/hle/service/apt/apt.cpp [hunk 9]
```diff
@@ -93,11 +101,176 @@ void GetSharedFont(Service::Interface* self) {
...
+    if (buffer_addr) {
+        u32 thread_id = 0;
+        u32 top_screen_color_per_pixel;
+        u32 bottom_screen_color_per_pixel;
+
+        GSP_GPU::FrameBufferUpdate* top_screen = GSP_GPU::GetFrameBufferInfo(thread_id, 0);
+        GSP_GPU::FrameBufferUpdate* bottom_screen = GSP_GPU::GetFrameBufferInfo(thread_id, 1);
+
+        capture_buffer_info.is_3d = 0; //not 3D
+        //main screen config
```
Comment (Subv) [inline src/core/hle/service/apt/apt.cpp]: Whitespace after all //

File: src/core/hle/service/apt/apt.cpp [hunk 10]
```diff
@@ -120,21 +293,30 @@ void GetLockHandle(Service::Interface* self) {
...
+    LOG_WARNING(Service_APT, "(STUBBED) called applet_pos=0x%08X", applet_pos);
+}
+
+void CountRegisteredApplet(Service::Interface* self) {
+    u32* cmd_buff = Kernel::GetCommandBuffer();
 
-    LOG_WARNING(Service_APT, "(STUBBED) called unk=0x%08X", unk);
+    cmd_buff[1] = RESULT_SUCCESS.raw; // No error
+    cmd_buff[2] = HLE::Applets::GetRegisteredAppletCount();
+    LOG_WARNING(Service_APT, "(STUBBED) called registered_applet_count=0x%08X", cmd_buff[2]);
```
Comment (Subv) [inline src/core/hle/service/apt/apt.cpp]: Is this still a stub?

File: src/core/hle/service/apt/apt.cpp [hunk 11]
```diff
@@ -272,36 +471,87 @@ void PrepareToStartApplication(Service::Interface* self) {
...
+    bool shell_open = true;
+
+    switch(static_cast<UtilityID>(utility_id)) {
+    /* void function(void)
+     * The utility SleepIfShellClosed deside the 3DS system whether in sleep state at short intervals (by check PTM::shell state?)
+     */
+    case UtilityID::SleepIfShellClosed:
+        shell_open = PTM::GetShellState();
+        if (!shell_open) {
+            LOG_WARNING(Service_APT, "(STUBBED) SleepIfShellClosed, The 3DS System should sleep now! :p");
```
Comment (Subv) [inline src/core/hle/service/apt/apt.cpp]:
```
(STUBBED) SleepIfShellClosed, The 3DS system should sleep now.
```

File: src/core/hle/service/apt/apt.cpp [hunk 12]
```diff
@@ -272,36 +471,87 @@ void PrepareToStartApplication(Service::Interface* self) {
...
+        } else {
+            LOG_WARNING(Service_APT, "(STUBBED) SleepIfShellClosed shell_open=%u",shell_open);
+        }
+        break;
+    case UtilityID::ClearPowerButtonState://not useful
```
Comment (Subv) [inline src/core/hle/service/apt/apt.cpp]: whitspace after all //

File: src/core/hle/service/apt/apt.cpp [hunk 13]
```diff
@@ -93,11 +101,176 @@ void GetSharedFont(Service::Interface* self) {
...
+
+void ReplySleepNotificationComplete(Service::Interface* self) {
+    u32* cmd_buff = Kernel::GetCommandBuffer();
+    u32 app_id = cmd_buff[1];
+
+    cmd_buff[1] = RESULT_SUCCESS.raw; // No error
+    LOG_WARNING(Service_APT, "(STUBBED) app_id=0x%08X", app_id);
+}
+
+void SendCaptureBufferInfo(Service::Interface* self) {
```
Comment (Subv) [inline src/core/hle/service/apt/apt.cpp]: SendCaptureBufferInfo should just copy the CaptureBufferInfo from cmd_buff[3] to a variable in our APT code. It should not regenerate a CaptureBufferInfo, that's what GSPGPU::ImportDisplayCaptureInfo is for.

http://3dbrew.org/wiki/APT:SendCaptureBufferInfo

**Comment by JamePeng** [comment]
In my opinion, the MessageParameter should be separated into four parts(Application, System Applet, Home Menu, Library Applet), which is more accurately for applet to deliver information and reducing the risk.

**Comment by neobrain** [comment]
Hi, this is neobot, using neobrain's account. I'm keeping an archive of versions of this PR:

2016-08-17T16:19:17Z: https://github.com/citra-emu/citra/compare/7b4dcacbb2006de6483e982b21956a8f3098aa1d...JamePeng:5cfd37225aac0bd34e24ca1110ec4363dc3cde22

**Comment by bunnei** [comment]
@JamePeng Sorry this has fallen into the void so long - can you please resolve merge conflicts?

**Comment by Subv** [comment]
I think this has been sitting around for enough time, if you still plan on continuing it, please open a new PR with the requested changes.

**Pull request closed without merging**